File: src/scanner.zig
const c = @import("./chunk.zig");
const Line = c.Line;
const std = @import("std");
const fmt = std.fmt;

pub const Token = struct {
    const Self = @This();
    type: TokenType,
    start: []const u8,
    line: Line,
};

pub const TokenType = enum {
    // Single-character tokens.
    LEFT_PAREN,
    RIGHT_PAREN,
    LEFT_BRACE,
    RIGHT_BRACE,
    COMMA,
    DOT,
    MINUS,
    PLUS,
    SEMICOLON,
    SLASH,
    STAR,
    // One or two character tokens.
    BANG,
    BANG_EQUAL,
    EQUAL,
    EQUAL_EQUAL,
    GREATER,
    GREATER_EQUAL,
    LESS,
    LESS_EQUAL,
    // Literals.
    IDENTIFIER,
    STRING,
    NUMBER,
    // Keywords.
    AND,
    CLASS,
    ELSE,
    FALSE,
    FOR,
    FUN,
    IF,
    NIL,
    OR,
    PRINT,
    RETURN,
    SUPER,
    THIS,
    TRUE,
    VAR,
    WHILE,

    ERROR,
    EOF,
};

pub const Scanner = struct {
    const Self = @This();
    start: []const u8,
    current: []const u8,
    line: Line,

    pub fn init(source: []const u8) !Self {
        return Self{
            .start = source,
            .current = source,
            .line = 1,
        };
    }

    pub fn scanToken(self: *Self) Token {
        self.skipWhitespace();
        self.start = self.current;

        if (self.isAtEnd())
            return self.makeToken(.EOF);

        const char = self.advance();

        if (isAlpha(char)) return self.identifier();
        if (isDigit(char)) return self.number();

        switch (char) {
            '(' => return self.makeToken(.LEFT_PAREN),
            ')' => return self.makeToken(.RIGHT_PAREN),
            '{' => return self.makeToken(.LEFT_BRACE),
            '}' => return self.makeToken(.RIGHT_BRACE),
            ';' => return self.makeToken(.SEMICOLON),
            ',' => return self.makeToken(.COMMA),
            '.' => return self.makeToken(.DOT),
            '-' => return self.makeToken(.MINUS),
            '+' => return self.makeToken(.PLUS),
            '/' => return self.makeToken(.SLASH),
            '*' => return self.makeToken(.STAR),
            '!' => {
                if (self.match('=')) {
                    return self.makeToken(.BANG_EQUAL);
                } else {
                    return self.makeToken(.BANG);
                }
            },
            '=' => {
                if (self.match('=')) {
                    return self.makeToken(.EQUAL_EQUAL);
                } else {
                    return self.makeToken(.EQUAL);
                }
            },
            '<' => {
                if (self.match('=')) {
                    return self.makeToken(.LESS_EQUAL);
                } else {
                    return self.makeToken(.LESS);
                }
            },
            '>' => {
                if (self.match('=')) {
                    return self.makeToken(.GREATER_EQUAL);
                } else {
                    return self.makeToken(.GREATER);
                }
            },
            '"' => return self.string(),

            else => return self.makeErrorToken("Unexpected character."),
        }

        return self.makeErrorToken("Unexpected character.");
    }

    fn makeToken(self: *Self, tokenType: TokenType) Token {
        const length = @ptrToInt(self.current.ptr) - @ptrToInt(self.start.ptr);
        return Token{
            .type = tokenType,
            .start = self.start[0..length],
            .line = self.line,
        };
    }

    fn makeErrorToken(self: *Self, message: []const u8) Token {
        return Token{
            .type = .ERROR,
            .start = message,
            .line = self.line,
        };
    }

    fn isAtEnd(self: *Self) bool {
        return self.current.len <= 0 or self.current[0] == 0;
    }

    fn advance(self: *Self) u8 {
        const result = self.current[0];
        self.current.ptr += 1;
        self.current.len -= 1;
        return result;
    }

    fn match(self: *Self, expected: u8) bool {
        if (self.isAtEnd())
            return false;
        if (self.current[0] != expected) return false;
        _ = self.advance();
        return true;
    }

    fn peek(self: *Self) u8 {
        if (self.isAtEnd())
            return 0;
        return self.current[0];
    }

    fn peekNext(self: *Self) u8 {
        if (self.isAtEnd() or self.current.len < 2)
            return 0;
        return self.current[1];
    }

    fn skipWhitespace(self: *Self) void {
        while (self.current.len > 0) {
            const char = self.peek();
            switch (char) {
                0xAA, ' ', '\r', '\t' => _ = self.advance(),
                '\n' => {
                    self.line += 1;
                    _ = self.advance();
                },
                '/' => {
                    if (self.peekNext() == '/') {
                        while (self.peek() != '\n' and !self.isAtEnd()) {
                            _ = self.advance();
                        }
                    } else {
                        return;
                    }
                },
                else => return,
            }
        }
    }

    fn checkKeyword(self: *Self, start: usize, length: usize, rest: []const u8, tokenType: TokenType) TokenType {
        if (@ptrToInt(self.current.ptr) - @ptrToInt(self.start.ptr) != start + length) // compare length
            return .IDENTIFIER;

        const source = self.start[start .. start + length];
        if (std.mem.eql(u8, source, rest)) // compare strings
            return tokenType;

        return .IDENTIFIER;
    }

    fn identifierType(self: *Self) TokenType {
        switch (self.start[0]) {
            'a' => return self.checkKeyword(1, 2, "nd", .AND),
            'c' => return self.checkKeyword(1, 4, "lass", .CLASS),
            'e' => return self.checkKeyword(1, 3, "lse", .ELSE),
            'f' => {
                if (@ptrToInt(self.current.ptr) - @ptrToInt(self.start.ptr) > 1) {
                    switch (self.start[1]) {
                        'a' => return self.checkKeyword(2, 3, "lse", .FALSE),
                        'o' => return self.checkKeyword(2, 1, "r", .FOR),
                        'u' => return self.checkKeyword(2, 1, "n", .FUN),
                        else => return .IDENTIFIER,
                    }
                }
            },
            'i' => return self.checkKeyword(1, 1, "f", .IF),
            'n' => return self.checkKeyword(1, 2, "il", .NIL),
            'o' => return self.checkKeyword(1, 1, "r", .OR),
            'p' => return self.checkKeyword(1, 4, "rint", .PRINT),
            'r' => return self.checkKeyword(1, 5, "eturn", .RETURN),
            's' => return self.checkKeyword(1, 4, "uper", .SUPER),
            't' => {
                if (@ptrToInt(self.current.ptr) - @ptrToInt(self.start.ptr) > 1) {
                    switch (self.start[1]) {
                        'h' => return self.checkKeyword(2, 2, "is", .THIS),
                        'r' => return self.checkKeyword(2, 2, "ue", .TRUE),
                        else => return .IDENTIFIER,
                    }
                }
            },
            'v' => return self.checkKeyword(1, 2, "ar", .VAR),
            'w' => return self.checkKeyword(1, 4, "hile", .WHILE),

            else => return .IDENTIFIER,
        }

        return .IDENTIFIER;
    }

    fn identifier(self: *Self) Token {
        while (isAlpha(self.peek()) or isDigit(self.peek())) _ = self.advance();
        return self.makeToken(self.identifierType());
    }

    fn string(self: *Self) Token {
        while (self.peek() != '"') {
            if (self.peek() == '\n') self.line += 1;
            _ = self.advance();
        }
        if (self.isAtEnd()) return self.makeErrorToken("Unterminated string.");
        _ = self.advance();
        return self.makeToken(.STRING);
    }

    fn number(self: *Self) Token {
        while (isDigit(self.peek())) _ = self.advance();
        if (self.peek() == '.' and isDigit(self.peekNext())) {
            _ = self.advance();
            while (isDigit(self.peek())) _ = self.advance();
        }
        return self.makeToken(.NUMBER);
    }
};

fn isAlpha(char: u8) bool {
    return (char >= 'a' and char <= 'z') or (char >= 'A' and char <= 'Z') or char == '_';
}

fn isDigit(char: u8) bool {
    return char >= '0' and char <= '9';
}

File: Keto/Scanner.cs
using System.Diagnostics.CodeAnalysis;

namespace Keto;

public class Scanner
{
    private readonly string _source;
    private int _startOffset;
    private int _currentOffset;
    private int _line = 1;

    private char Current
    {
        get
        {
            if (_currentOffset >= _source.Length)
                return '\0';
            return _source[_currentOffset];
        }
    }

    private char Next
    {
        get
        {
            if (IsAtEnd())
                return '\0';
            return _source[_currentOffset + 1];
        }
    }

    public Scanner(string source)
    {
        _source = source;
    }

    public Token ScanToken()
    {
        SkipWhitespace();
        _startOffset = _currentOffset;
        if (IsAtEnd())
            return MakeToken(TokenType.Eof);

        var character = Advance();
        if (IsDigit(character))
            return MakeNumber();
        if (IsAlpha(character))
            return MakeIdentifier();

        switch (character)
        {
            case '(': return MakeToken(TokenType.LeftParen);
            case ')': return MakeToken(TokenType.RightParen);
            case '{': return MakeToken(TokenType.LeftBrace);
            case '}': return MakeToken(TokenType.RightBrace);
            case ';': return MakeToken(TokenType.Semicolon);
            case ',': return MakeToken(TokenType.Comma);
            case '.': return MakeToken(TokenType.Dot);
            case '-': return MakeToken(TokenType.Minus);
            case '+': return MakeToken(TokenType.Plus);
            case '/': return MakeToken(TokenType.Slash);
            case '*': return MakeToken(TokenType.Star);
            case '!':
                return MakeToken(Match('=') ? TokenType.BangEqual : TokenType.Bang);
            case '=':
                return MakeToken(Match('=') ? TokenType.EqualEqual : TokenType.Equal);
            case '<':
                return MakeToken(Match('=') ? TokenType.LessEqual : TokenType.Less);
            case '>':
                return MakeToken(Match('=') ? TokenType.GreaterEqual : TokenType.Greater);
            case '"':
                return MakeString();
        }

        return ErrorToken("Unexpected character.");
    }

    private char Advance()
    {
        _currentOffset++;
        return _source[_currentOffset - 1];
    }

    private bool Match(char expected)
    {
        if (IsAtEnd())
            return false;
        if (Current != expected)
            return false;
        _currentOffset++;
        return true;
    }

    private bool IsAtEnd()
    {
        return Current == '\0';
    }

    private bool IsDigit(char c)
    {
        return c >= '0' && c <= '9';
    }

    private bool IsAlpha(char c)
    {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    private Token MakeToken(TokenType type)
    {
        var length = _currentOffset - _startOffset;
        return new Token
        {
            Type = type,
            StartOffset = _startOffset,
            Length = length,
            Line = _line,
            Value = _source.Substring(_startOffset, length)
        };
    }

    private Token ErrorToken(string message)
    {
        return new Token
        {
            Type = TokenType.Error,
            StartOffset = 0,
            Length = message.Length,
            Line = _line,
            Value = message
        };
    }

    private Token MakeString()
    {
        while (Current != '"' && !IsAtEnd())
        {
            if (Current == '\n')
                _line++;
            Advance();
        }

        if (IsAtEnd())
            return ErrorToken("Unterminated string.");
        Advance();
        return MakeToken(TokenType.String);
    }

    private Token MakeNumber()
    {
        while (IsDigit(Current))
            Advance();

        // Look for fractional part
        if (Current == '.' && IsDigit(Next))
        {
            Advance(); // consume the '.'
            while (IsDigit(Current))
                Advance();
        }

        return MakeToken(TokenType.Number);
    }

    private Token MakeIdentifier()
    {
        while (IsAlpha(Current) || IsDigit(Current))
            Advance();
        return MakeToken(IdentifierType());
    }

    [SuppressMessage("ReSharper", "StringLiteralTypo")]
    private TokenType IdentifierType()
    {
        switch (_source[_startOffset])
        {
            case 'a': return CheckKeyword(1, 2, "nd", TokenType.And);
            case 'c': return CheckKeyword(1, 4, "lass", TokenType.Class);
            case 'e': return CheckKeyword(1, 3, "lse", TokenType.Else);
            case 'i': return CheckKeyword(1, 1, "f", TokenType.If);
            case 'n': return CheckKeyword(1, 2, "il", TokenType.Nil);
            case 'o': return CheckKeyword(1, 1, "r", TokenType.Or);
            case 'p': return CheckKeyword(1, 4, "rint", TokenType.Print);
            case 'r': return CheckKeyword(1, 5, "eturn", TokenType.Return);
            case 's': return CheckKeyword(1, 4, "uper", TokenType.Super);
            case 'v': return CheckKeyword(1, 2, "ar", TokenType.Var);
            case 'w': return CheckKeyword(1, 4, "hile", TokenType.While);

            case 'f':
                if (_currentOffset - _startOffset > 1)
                    switch (_source[_startOffset + 1])
                    {
                        case 'a': return CheckKeyword(2, 3, "lse", TokenType.False);
                        case 'o': return CheckKeyword(2, 1, "r", TokenType.For);
                        case 'u': return CheckKeyword(2, 1, "n", TokenType.Fun);
                    }

                break;
            case 't':
                if (_currentOffset - _startOffset > 1)
                    switch (_source[_startOffset + 1])
                    {
                        case 'h': return CheckKeyword(2, 2, "is", TokenType.This);
                        case 'r': return CheckKeyword(2, 2, "ue", TokenType.True);
                    }

                break;
        }

        return TokenType.Identifier;
    }

    private TokenType CheckKeyword(int start, int length, string rest, TokenType type)
    {
        if (_currentOffset - _startOffset == start + length && _source.Substring(_startOffset + start, length) == rest)
            return type;
        return TokenType.Identifier;
    }

    private void SkipWhitespace()
    {
        while (true)
        {
            char c = Current;
            switch (c)
            {
                case ' ':
                case '\r':
                case '\t':
                    Advance();
                    break;
                case '\n':
                    _line++;
                    Advance();
                    break;
                case '/':
                    if (Next == '/')
                        while (Current != '\n' && !IsAtEnd())
                            Advance();
                    else
                        return;
                    break;
                default:
                    return;
            }
        }
    }
}